docs/mem-management/man-vs-auto-and-gc.md

```diff
@@ -0,0 +1,132 @@
+---
+title: Manual vs. Automatic Memory Management and The Garbage Collector
+parent: Memory Management
+nav_order: 2
+---
+
+# Memory management in other systems programming languages
+
+Now that we understand the necessity of dynamically allocated memory, let's compare manual and automatic ways of management.
+
+C, which inspired the design of many of the languages that came after it, had no automatic memory management out of the box. You could make your own GC library or use a publicly available one, but it did not fit well with the design philosophy of the language: giving full responsability to the programmer, with minimal abstractions. C++ mostly followed in the footsteps of C in this regard, until C++11, when [smart pointers](https://docs.microsoft.com/en-us/cpp/cpp/smart-pointers-modern-cpp?view=msvc-170) were introduced, simplifying the use of dynamic memory. In contrast, Java has a complex automatic memory management system, based on a mark-and-sweep type of garbage collection (GC).
+
+# The Garbage Collector
+
+Ali Çehreli, in his Programming in D book, gives a good introduction to how the GC works at a high-level:
+
+```
+The dynamic variables that are used in D programs live on memory blocks that are owned by the garbage collector (GC). When the lifetime of a variable ends (i.e. it's no longer being used), that variable is subject to being finalized according to an algorithm that is executed by the GC. If nothing else needs the memory location containing the variable, the memory may be reclaimed to be used for other variables. This algorithm is called garbage collection and an execution of the algorithm is called a garbage collection cycle.
+```
+
+# Memory management in D
+
+The [D language specification](https://dlang.org/spec/garbage.html) offers a good insight into how the D's GC works, here we will present only the most important aspects of it. D makes extensive use of GC, even if it might not be obvious at a first glance. To make efficient use of it, we must understand when and how the GC does it's job. Exceptions, strings, dynamic and associative arrays are some of the built-in types that make use of GC-allocated memory. Let's look at an example, using the append operator:
+
+```d
+void main() {
+    int[] ints;
+    foreach(i; 0..100) {
+        ints ~= i;
+    }
+}
+```
+
+Appending elements to a dynamic array in this manner will trigger GC allocations when the capacity of the array is insufficient. You can see this by compiling the above code with the option ```-vgc```. Even better, we can use the [capacity](https://dlang.org/phobos/object.html#.capacity) property of the array to inspect the capacity at any time.
```

Comment:
Add a refresher on the differences between `size` and `capacity`. This will make the explanation easier to understand